### ISSUE TITLE
Missing '$IdentityOptG_new'

### ISSUE THREAD

**Issue by plajjan** [actonlang/acton#1713]
### Acton Version

0.19.2.20240121.17.6.48

### Steps to Reproduce and Observed Behavior

Try to check for None on a object member of optional type and we get an error.

```
class Foo(object):
    a: ?str
    b: ?str
    def __init__(self, a, b):
        self.a = a
        self.b = b


actor main(env):
    f = Foo("a", "b")
    if f.a is not None:
        pass

    env.exit(0)
```

we get this error:
```
/tmp/actonc-4dec71409d4b9763/out/types/test_none_in_class.c:107:36: error: call to undeclared function '$IdentityOptG_new'; ISO C99 and later do not support implicit function declarations
```

### Expected Behavior

It should work!

**Comment by plajjan** [comment]
The workaround to this is fairly simple... but I think this should just work, right? @nordlander @sydow

**Comment by nordlander** [comment]
I agree in principle, but there's a catch: we want `?A` to be a base class of every class `A`, even those not yet defined! That requirement fits very badly with our class extension mechanism, which is why `?A` has been treated as a rather odd (and buggy) special construct up to now.

Luckily, our expectations on `?A` blend rather well with the idea of _union types_, where `?A` can be defined as just a shorthand for the union `None|A`. And union types can also be made to support protocol extensions in a quite flexible way, based on the rule that `T1|T2` implements protocol `P` if and only if both `T1`and `T2` implement `P`.

For these reasons I'm inclined to see union types as the long term solution to whatever problems we currently have with the optional type. Especially so if we have a workaround that we can live with for a while.

**Comment by plajjan** [comment]
So the workaround is to introduce a local variable from the attribute of the object we want to inspect
```
actor main(env):
    f = Foo("a", "b")
    a = f.a
    if a is not None:
        pass

    env.exit(0)
```

I don't think it's urgent to fix this, since it's been a problem since all of eternity and we have this simple workaround... but it's annoying and I didn't want to loose track of it. Thus this issue. I think it can wait for unions, which should be just after integer subtyping :)

**Comment by plajjan** [comment]
@nordlander we briefly talked about this issue today. The previous comments here indicate a dependency on unions. While I don't remember your exact words, I don't think it really aligned with that. Do you think there is a natural path forward for this without unions? I think you also mentioned a conceptual design choice in our call while the comments here about unions don't really seem like that.

Can you please jot down a summary characterization of the issue and our options for doing something about this?

**Comment by nordlander** [comment]
Here's my concrete proposal:

1. Define `<expr> is None` and  `<expr> is not None` as special syntactic forms that do pattern-matching against the `None` constructor _for all forms of expressions_.
2. Remove the pattern-matching and narrowing behavior of `<expr> != None` so that `!=` on an optional type becomes a standard equality test. Ditto for `<expr> == None`.
3. Keep the type narrowing feature of `<variable> is not None` (to be extended later to handle more expression forms). 

I've implemented this proposal and verified that it solves the core problem of this issue. It also works with all our existing code repos (after systematic replacement of `!= None` and `== None`). But the really cool thing is that it's also compatible with the general pattern-matching extension we're planning, such that
```
<expr> is None
```
just becomes a special case of
```
<expr> is <pattern>
```
and
```
if <expr> is <pattern>:
    <do_someting>
else:
    <do_something_else>
```
can be seen as a syntactic sugar for
```
match <expr>:
    case <pattern>:
        <do_something>
    case _:
        <do_something_else>
```
(and vice versa for `is not` tests).

The seemingly ambiguous overlap with the more standard use of `is` as an identity comparison operator is actually benign, since comparison with a constructor pattern interpreted as an expression would either be illegal (if an immutable type is constructed) or meaningless (since identity comparison with a newly constructed object would always return `False`).

So I think this proposal will place testing against `None` on some firmer ground, while also erasing some gaping holes in our implementation. I'll go ahead and push this soon (together with boring updates of the affected repos) if nobody shouts!
